feat: Add CloudWatch logs subscription handling support to awslambdareceiver

Kavindu-Dodan opened the pull request:
#### Description

This PR adds CloudWatch logs Lambda subscription handling support to `awslambdareceiver`.

<img width="514" height="222" alt="image" src="https://github.com/user-attachments/assets/6b170e24-32ed-4f05-a985-d134c6c9b26c" />

With this change, the receiver can now be deployed as a Lambda and support handling either of the signals: S3 triggers or CloudWatch subscription logs.

#### Link to tracking issue

Completes part of #43504 

#### Testing

I have added unit tests to cover most of the new functionality. One of them (`TestHandleCloudwatchLogEvent`) provides end to end validation of a CloudWatch event to plogs conversion. 

#### Documentation

Existing documentation already covered some aspects. However, I have attempted to uplift them along with code level comments.

Kavindu-Dodan committed:
feat: improve lambda execution detection

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan committed:
feat: add cloudwatch log handling support

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

# Conflicts:
#	receiver/awslambdareceiver/go.sum

# Conflicts:
#	receiver/awslambdareceiver/go.sum

# Conflicts:
#	receiver/awslambdareceiver/go.sum

axw commented on receiver/awslambdareceiver/config.go:
```diff
@@ -15,10 +15,10 @@ const (
 	//
-	// If you have objects with multiple different encodings to handle, you should deploy
-	// separate Lambda functions with different configurations.
+	// If unspecified, the receiver falls back to work with CloudWatch Log subscription encoding extension.
```
I know we went back and forth on this internally, but I'd really like to keep these things independent. Can we maintain the existing doc comment please? It's confusing to talk about CloudWatch logs in the same breath as S3 encoding - they're totally separate events.

Kavindu-Dodan replied:
I am exploring the option described here - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562#issuecomment-3587411621 . Let me know your view on this.

axw reviewed (commented):
Code looks fine, but I'd like to revisit the config. Implicitly making it CloudWatch-based because the S3 encoding isn't configured is a bit unintuitive.

MichaelKatsoulis commented:
Reading your suggestions, I agree that keeping the configuration for s3 and cloudwatch_logs separate is the cleaner approach.

> How would that encoding extension be used for CloudWatch Logs? ... I think it would make sense to use it for decoding the message field.

Yes, exactly. If `message_encoding` is provided, we should use it to decode the message field of the CloudWatch Log Events.

However, there is a significant technical challenge here regarding the input format expectations.

1. Mismatch
The CloudWatch Subscription Filter sends us a JSON payload where logEvents is an array of individual 
messages:
```
"logEvents": [
        {
            "id": "31953106606966983378809025079804211143289615424298221568",
            "timestamp": 1432826855000,
            "message": "{\"eventVersion\":\"1.03\",\"userIdentity\":{\"type\":\"Root\"}"
        },
        {
            "id": "31953106606966983378809025079804211143289615424298221569",
            "timestamp": 1432826855000,
            "message": "{\"eventVersion\":\"1.03\",\"userIdentity\":{\"type\":\"Root\"}"
        },
        {
            "id": "31953106606966983378809025079804211143289615424298221570",
            "timestamp": 1432826855000,
            "message": "{\"eventVersion\":\"1.03\",\"userIdentity\":{\"type\":\"Root\"}"
        }
    ]
```
However, the awslogs encoding extension (e.g., for CloudTrail) expects a specific root-level structure that mirrors the original S3 file format:
```
{ 
   "Records": [        
       { "eventVersion": "1.08"... },        
       { "eventVersion": "1.08"... }    
   ]
}
```
2. Transformation Requirement
We cannot simply stream the message fields into the encoding extension.
For VPC Flow Logs: Concatenating messages with \n might work (as they are line-delimited).
For CloudTrail: We would need to explicitly construct a valid JSON wrapper {"Records": [ ... ]} around the concatenated messages before passing them to the unmarshaler.
3. Efficiency Concern
Ideally, if `message_encoding` is set, we should indeed bypass the full subscriptionFilter unmarshaler to avoid the overhead of `Bytes -> PLogs (SubFilter) -> Bytes (Re-encoded) -> PLogs (Target)`.  We can parse the raw CloudWatch JSON, extract the message strings directly, apply the necessary wrapping (e.g. {"Records":[ + joined messages + ]}), and feed that byte stream to the target unmarshaler.

I agree this implementation detail can be tackled in a follow-up PR, but it's important to note that supporting message_encoding might require the receiver to have some logic to "adapt" the stream (delimiter/wrapping) based on the target encoding type.

Kavindu-Dodan committed:
feat: enhance receiver & handler behaviour

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan commented on receiver/awslambdareceiver/README.md:
```diff
@@ -54,9 +54,9 @@ The `awslambdareceiver` operates as follows:
-| `s3_encoding`         | Name of the encoding extension to use for S3 objects                                                                                       | "awslogs_encoding" | Optional |
+| Name          | Description                                                                                                                                                                                        | Required |
+|:--------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|
+| `s3_encoding` | Name of the encoding extension to use for S3 objects. For example, `awslogs_encoding` to use AWS logs encoding extensions. When unspecified, falls back to CloudWatch subscription filter handling | Optional |
```
I will resolve this and let's continue this through main comments

Kavindu-Dodan commented on receiver/awslambdareceiver/receiver.go:
```diff
@@ -164,9 +166,14 @@ func newLogsHandler(
+			return nil, err
+		}
+
+		return newCWLogsSubscriptionHandler(set.Logger, unmarshaler.UnmarshalLogs, next.ConsumeLogs), nil
```
I will resolve this and let's continue this through main comments

Kavindu-Dodan commented:
@axw @MichaelKatsoulis thanks for the feedback. Let me clarify my view and let's see if we can come to conclusion on this.

First, let's break down receiver operation modes. The lambda receiver can operate for both logs & metrics. Each operation mode can then have sub-trigger modes.  Currently, we support the following,

- Logs : Trigger from S3 & Trigger from CloudWatch 
- Metrics : Trigger from S3


The detection of trigger type is easy and is already done by `detectTriggerType` implementation. Once the trigger type is detected, we have well-defined event formats to extract payload that we must process. For example, for the S3 trigger, the event format is known [^1] and that's what we use to obtain the content of the S3 object. 

Once we obtain the content, that's where we need the help of encoding extension. So if the user is forwarding VPC flow logs to S3, then user must use `awslogs_encoding` with `vpcflow` format

```yaml
receivers:
  awslambda:
    encoding_extension: awslogs_encoding

extensions:
  awslogs_encoding:
    format: vpcflow
    vpcflow:
      file_format: plain-tex
```

> How would that encoding extension be used for CloudWatch Logs? They have a specific structure -- they must be decoded in the CloudWatch log subscription filter format.

If we accept the proposed change in this PR, then the user must define `awslogs_encoding` with `cloudwatch` format, 

```yaml
receivers:
  awslambda:
    encoding_extension: awslogs_encoding

extensions:
  awslogs_encoding:
    format: cloudwatch
```

However, this is where we sort of misuse the CloudWatch subscription filter. Because in the implementation [^2], we are just reading the content of the message and set it to the scope log log record. In other words, if we apply the same usage of S3 event trigger, we are performing content download of the S3 object through an encoding extension.

Because of this, we have the encoder chaining need, and the complexity comes through this is explained by @MichaelKatsoulis at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562#issuecomment-3589063041

___

### Improving CloudWatch subscription trigger mode

As the next step I have a suggestion. For the CloudWatch trigger we must move message extraction into the Lambda receiver. Once we have the message content, we can choose two paths,

1 Decode the message further using the defined `encoding_extension`
2 Define operation mode where CW logs are converted as is

The first path fixes the encoding chaining issue. And we can use improved encoding extensions  (see note [^3])  to decode CW logs.

The second path exactly matches what we have now. Defining this operation mode needs some refactoring and can be specific to CW event type. 

@axw @MichaelKatsoulis let me know your thoughts. Also check the current state where commit https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562/commits/fbd585f44c6bc7d4d318afef53a47e7d90eb53bf contains changes towards this direction. 


[^1]: https://docs.aws.amazon.com/lambda/latest/dg/with-s3.html
[^2]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/10c1e40324f32587e6c8c0b1bcced617486c6cb1/extension/encoding/awslogsencodingextension/internal/unmarshaler/subscription-filter/unmarshaler.go
[^3]: We still need improvements as encoding extensions are specialized to handle S3 triggers. For example, VPC flow logs expect the first line to contain field names -  https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/10c1e40324f32587e6c8c0b1bcced617486c6cb1/extension/encoding/awslogsencodingextension/internal/unmarshaler/vpc-flow-log/unmarshaler.go#L96-L99

axw commented:
@Kavindu-Dodan no argument from me about supporting encoding extensions for CloudWatch logs. I still think they should be kept independent as I described above.

> However, the awslogs encoding extension (e.g., for CloudTrail) expects a specific root-level structure that mirrors the original S3 file format:

@MichaelKatsoulis ah, right. We really want to process the whole log file at once, since it has the metadata in there too. Maybe it's not worth trying to parse AWS service logs from CloudWatch specifically, but encoding extension support could still be useful. Anyway, as you said, we can consider those details later.

Kavindu-Dodan committed:
add built-in unmarshaling

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan committed:
go mod tidy

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

# Conflicts:
#	receiver/awslambdareceiver/go.mod
#	receiver/awslambdareceiver/go.sum

# Conflicts:
#	receiver/awslambdareceiver/go.mod
#	receiver/awslambdareceiver/go.sum

# Conflicts:
#	receiver/awslambdareceiver/go.mod
#	receiver/awslambdareceiver/go.sum

Kavindu-Dodan committed:
parse string as is for logs

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan committed:
rename config option, enahnce metrics handling and improve cw logs mode

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan commented:
>> Just double checking: for CloudWatch Logs, the encoding will only apply to the message field right?

> Based on your response: the answer is no, it applies to the entire payload.

Sorry yes I missed this. Yes for CW trigger encoding will apply to entire payload. The payload internally contain the actual log messages. 

Regarding the approach, I still needs few clarifications, so bare with me :) 

> By default (no encoding specified), we create an OTLP log record for each CloudWatch log event within the payload. The log event message will be used exactly as-is, and users can use transforms to further process.

Does this mean we will create one `plog.Logs` per each message inside CW trigger ? For example,  if we receive following,

```json

{
  "owner": "123456789012",
  "messageType": "DATA_MESSAGE",
  "logEvents": [
    {
      "id": "evt-001",
      "timestamp": 1732638456000,
      "message": "one"
    },
    {
      "id": "evt-002",
      "timestamp": 1732638469000,
      "message": "two"
    }
  ]
}
```

We will emit two plogs for consumer ? 

With your explanation on usage, it makes sense. But just want to cross check if this is what you mean :)

> If cloudwatch::encoding is specified, it will apply to the entire Lambda event payload, and not to each message.

Regarding this and combining with the view `but I'd still like to consider make the encoding config event/handler-specific (i.e. s3 encoding, cloudwatch encoding)`

Can we extract this to a specific encoding extension ? I mean the behaviour where the extension extracts each message into logrecord body  (similar to test output at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562/changes#diff-77f3be94371796504a88b0c9d8dad6ee473b21c473967c55cd85d4c4a765aea9) ? This may be done by upgrading `subscription-filter` ([link](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/encoding/awslogsencodingextension/internal/unmarshaler/subscription-filter)) unmarshaler, which is obsolete at its current state, as we have first level processing at Lambda receiver. With the upgraded unmarshaling, we can add the full message processing to there.

axw commented:
> Sorry, I don't follow...

Maybe you thought I meant that `cloudwatch::encoding` would apply to the individual log events still? To clarify, I'm on board with having the encoding apply to the whole subscription filter event. I'm just saying I'd rather have a separate `encoding` config for S3 and CloudWatch.

Kavindu-Dodan committed:
review change: add trigger specific configuration, improve documentation

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan commented:
@axw see commit https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562/changes/b4b445909e39acb596b7b52549461e6370dbcbf2

I have updated the PR with our agreed configuration approach. Please have a review when you have time :)

axw commented on receiver/awslambdareceiver/README.md:
```diff
@@ -36,34 +32,61 @@ The `awslambdareceiver` operates as follows:
+
+> [!IMPORTANT]  
+> Currently, metrics mode will always use `awscloudwatchmetricstreams_encoding` extension regardless of the `encoding` parameter set in the receiver configuration.
+> Please make sure ingesting metrics can be decoded using this extension.
```
Should we error out if `encoding` is set then, when the receiver is used in a metrics pipeline? Maybe in the future we can allow it to be set to decode metrics stored in some other format, e.g. OTLP protobuf

axw replied:
I agree

axw commented on receiver/awslambdareceiver/handler.go:
```diff
@@ -115,6 +184,129 @@ func (*s3Handler[T]) parseEvent(raw json.RawMessage) (event events.S3EventRecord
+	return logs
+}
+
+func bytesToPlogs(data []byte, event events.S3EventRecord) any {
```
```suggestion
func bytesToPlogs(data []byte, event events.S3EventRecord) plog.Logs {
```

axw commented on receiver/awslambdareceiver/handler.go:
```diff
@@ -77,12 +135,23 @@ func (s *s3Handler[T]) handle(ctx context.Context, event json.RawMessage) error
-		return fmt.Errorf("failed to unmarshal logs: %w", err)
+	var data T
+	if s.s3Unmarshaler == nil {
+		switch any(s.emitType).(type) {
```
This is a bit hacky, type-asserting on a zero value when we should already know from s3Handler's type parameter. Rather than doing this, can we just arrange for `s3Unmarshaler` to always be non-nil?

When creating the s3Handler for logs and an encoding hasn't been specified, we can create an unmarshaler that uses bytesToPlogs.

axw replied:
Depending on the outcome of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562#pullrequestreview-3570161686, maybe we could remove all default unmarshalling code from the receiver, and:
 - default to using the "text_encoding" extension for S3 logs
 - default to using the "awscloudwatchmetrics_encoding" extension for S3 metrics
 - default to using the "awslogs_encoding" extension with `format: cloudwatch` for CloudWatch Logs

axw reviewed (commented):
> for s3 do you mean that we would put the whole s3 file content in one log record? Is this because it is hard to detect the right delimiter to split the contents into different records (e.g line by line)? Then it makes sense.
> If the file is consisted of json records then user would choose the json encoding extension.

@MichaelKatsoulis I did mean the whole s3 file (which IIANM is what we're doing in this PR), but now that you ask this I'm second-guessing.

I mostly wanted to separate the configs (which is done), and I'm +1 on having a sensible default when encoding isn't specified. For S3 logs, I think it would be fine for the default to producing a record per line, where the whole line is used as the body as-is -- I don't recall if there was a specific reason to not do that.

@MichaelKatsoulis @Kavindu-Dodan WDYT?

Kavindu-Dodan committed:
review changes: always set unmarshaler, metrics defaults, update readme

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

# Conflicts:
#	receiver/awslambdareceiver/handler_test.go

Kavindu-Dodan commented on receiver/awslambdareceiver/handler.go:
```diff
@@ -77,12 +135,23 @@ func (s *s3Handler[T]) handle(ctx context.Context, event json.RawMessage) error
+	if s.s3Unmarshaler == nil {
+		switch any(s.emitType).(type) {
+		case plog.Logs:
+			data = bytesToPlogs(body, parsedEvent).(T)
```
We no longer have this with the latest refactoring. I had to use this when `s3Unmarshaler` could be nil.

axw commented on receiver/awslambdareceiver/handler.go:
```diff
@@ -115,13 +161,142 @@ func (*s3Handler[T]) parseEvent(raw json.RawMessage) (event events.S3EventRecord
+	sl.Scope().SetName(metadata.ScopeName)
+
+	lr := sl.LogRecords().AppendEmpty()
+	lr.Attributes().PutInt("content.length", int64(len(data)))
```
```suggestion
```

Doesn't seem to add any value - can't we easily determine the length from the body?

Kavindu-Dodan replied:
I originally thought it might help with some efforts like streaming as S3 bound files can be large, but given that we add everything to plog, yes it doesn't add value. So will remove it.

axw commented on receiver/awslambdareceiver/README.md:
```diff
@@ -36,34 +32,68 @@ The `awslambdareceiver` operates as follows:
+Consider following notes on default behaviors:
+
+- When `s3:encoding` is not specified, the receiver defaults to preserving the S3 object content as-is for logs.
+  - However, receiver attempts to detect `strings` vs `non-strings` where `Body` will be set to `string` or `byte ` accordingly.
```
```suggestion
  - The log record's `Body` field will be a string type where the S3 object content is valid UTF-8, and otherwise will be a byte array.
```

"Attempts to detect" leaves some ambiguity, I think in this case we can be very explicit about how data is mapped since the logic is very simple.

axw reviewed (commented):
Thanks for all the back-and-forth @Kavindu-Dodan, just a few minor comments, otherwise LGTM.

BTW I'm back to thinking we should keep the default S3 logs decoding as you have it now. Users can always configure the receiver to use `text_encoding` if they want to; but they cannot configure the receiver to store the content as-is if it defaults to `text_encoding`.

Kavindu-Dodan committed:
review changes: readme, content length & type T

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

Kavindu-Dodan committed:
Update receiver/awslambdareceiver/README.md

Co-authored-by: Andrew Wilkins <axwalk@gmail.com>

Kavindu-Dodan commented on receiver/awslambdareceiver/handler.go:
```diff
@@ -4,50 +4,108 @@
 	logger        *zap.Logger
 	s3Unmarshaler unmarshalFunc[T]
 	consumer      s3EventConsumerFunc[T]
+	emitType      T
```
Yup, we no longer needs this (at least for now). So removed with https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44562/changes/0ceaa15fbe91faf16e9ca3ee26f342d4e83966bc

Kavindu-Dodan commented:
> Thanks for all the back-and-forth @Kavindu-Dodan, just a few minor comments, otherwise LGTM.
> 
> BTW I'm back to thinking we should keep the default S3 logs decoding as you have it now. Users can always configure the receiver to use `text_encoding` if they want to; but they cannot configure the receiver to store the content as-is if it defaults to `text_encoding`.

Thanks @axw for the detailed review. Good point on default encoding, so let's keep the current as-is default and move forward. We can improve this behaviour in future if there's demand for changes.

cc @MichaelKatsoulis may be a final review when you have time ?

MichaelKatsoulis reviewed (approved):
Thank you @Kavindu-Dodan .

This looks great!

Kavindu-Dodan committed:
make tests semconv independent

Signed-off-by: Kavindu Dodanduwa <kavindu.dodanduwa@elastic.co>

axw committed:
Update receiver/awslambdareceiver/go.mod
